lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -169,142 +194,235 @@ struct ProcedureOpConversion : public OpConversionPattern<ProcedureOp> {
-          Value posedge =
-              rewriter.create<comb::AndOp>(loc, oldValues[i], notCurrVal);
-          disjuncts.push_back(posedge);
-        }
+      Value trueVal = rewriter.create<hw::ConstantOp>(loc, APInt(1, 1));
+
+      if (edge == Edge::PosEdge || edge == Edge::BothEdges) {
+        Value notOldVal =
+            rewriter.create<comb::XorOp>(loc, before, trueVal, true);
+        Value posedge =
+            rewriter.create<comb::AndOp>(loc, notOldVal, after, true);
+        disjuncts.push_back(posedge);
       }
 
-      Value isValid = rewriter.create<comb::OrOp>(loc, disjuncts, false);
-      rewriter.create<cf::CondBranchOp>(loc, isValid, &op.getBody().front(),
-                                        wait);
-    }
+      if (edge == Edge::NegEdge || edge == Edge::BothEdges) {
+        Value notCurrVal =
+            rewriter.create<comb::XorOp>(loc, after, trueVal, true);
+        Value posedge =
+            rewriter.create<comb::AndOp>(loc, before, notCurrVal, true);
+        disjuncts.push_back(posedge);
+      }
 
-    for (auto event : events)
-      rewriter.eraseOp(event);
+      return rewriter.createOrFold<comb::OrOp>(loc, disjuncts, true);
+    };
 
-    rewriter.inlineRegionBefore(op.getBody(), procOp.getBody(),
-                                procOp.getBody().end());
+    // Convert all `detect_event` ops into a check for the corresponding event
+    // between the value before and after the `llhd.wait`. The "before" value
+    // has been collected into `valuesBefore` in the "wait" block; the "after"
+    // value corresponds to the detect op's input.
+    SmallVector<Value> triggers;
+    for (auto [detectOp, before] : llvm::zip(detectOps, valuesBefore)) {
+      rewriter.setInsertionPoint(detectOp);
+      auto trigger =
+          computeTrigger(before, detectOp.getInput(), detectOp.getEdge());
+      if (detectOp.getCondition()) {
+        auto condition = typeConverter->materializeTargetConversion(
+            rewriter, loc, rewriter.getI1Type(), detectOp.getCondition());
+        trigger = rewriter.create<comb::AndOp>(loc, trigger, condition, true);
+      }
+      triggers.push_back(trigger);
+      rewriter.eraseOp(detectOp);
+    }
 
-    for (auto returnOp : procOp.getOps<ReturnOp>()) {
-      rewriter.setInsertionPoint(returnOp);
-      rewriter.create<cf::BranchOp>(loc, wait);
-      rewriter.eraseOp(returnOp);
+    // If any `detect_event` op detected an event, branch to the "resume" block
+    // which contains any code after the `wait_event` op. If no events were
+    // detected, branch back to the "wait" block to wait for the next change on
+    // the interesting signals.
+    rewriter.setInsertionPointToEnd(checkBlock);
+    if (!triggers.empty()) {
+      auto triggered = rewriter.createOrFold<comb::OrOp>(loc, triggers, true);
+      rewriter.create<cf::CondBranchOp>(loc, triggered, resumeBlock, waitBlock);
+    } else {
+      rewriter.create<cf::BranchOp>(loc, waitBlock);
```

Comment:
In this case, the resume block would have no predecessors. Do you think it's save to leave this to the conversion framework? We don't have a test for this case to be sure.

Reply:
Good point, I have added the pattern I've seen failing before to the test with an empty `wait_event`. It looks like the dialect conversion uses the dominance information collected _before_ the conversion to iterate through ops -- looks like making a block unreachable doesn't exclude it from the conversion.

Reply:
Good point, I have added the pattern I've seen failing before to the test with an empty `wait_event`. It looks like the dialect conversion uses the dominance information collected _before_ the conversion to iterate through ops -- looks like making a block unreachable doesn't exclude it from the conversion.